bundles/org.openhab.binding.lcn/src/main/java/org/openhab/binding/lcn/internal/connection/AbstractState.java

```diff
@@ -16,55 +16,61 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.ScheduledFuture;
+import java.util.function.Function;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
 
 /**
- * Base class for usage for states with {@link StateMachine}.
+ * Base class for all states used with {@link AbstractStateMachine}.
+ *
+ * @param <T> type of the state machine implementation
+ * @param <U> type of the state implementation
  *
  * @author Fabian Wolter - Initial Contribution
  */
 @NonNullByDefault
-public abstract class AbstractState {
+public abstract class AbstractState<T extends AbstractStateMachine<T, U>, U extends AbstractState<T, U>> {
```

Comment:
I tried several approaches, but couldn't get rid of the generics. Any suggestions?

Reply:
Not sure what to say. As an example, this code compiles just fine:
```

    public interface State {

    }

    public interface StateMachine {
        public void nextState(Function<StateMachine, State> stateFactory);
    }

    public class StateImpl implements State {

        private StateMachine context;

        public StateImpl(StateMachine context) {
            this.context = context;
        }

        public void changeState() {
            context.nextState(StateImpl::new);
        }
    }
```

Reply:
The problems arise when storing data in the context like the connection object. These data cannot be accessed from within `StateImpl`, because the reference to the context is of the interface's type. I used generics to be able to access the concrete type.

The state machine and state abstraction layer could be removed (and so the generics), but then the code wouldn't be reusable.

Another approach is to replace the generics by casts. If I see correctly only a single cast would be necessary.

As always, there are several solutions with their pros and cons. Maybe I hadn't *the* bright idea, yet.